### PR TITLE
Generation of measureLog.json all the pcr event log

### PR DESCRIPTION
**Pull request by mjeyarax**
Included generation of measureLog.json with uefi, TXT, 501 and application events.

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -42,3 +43,57 @@ test:
       - "out/cover.html"      
       - "out/tagent"
     expire_in: 1 year
+
```
Comment (kwtj43) [inline .gitlab-ci.yml]: Is this going to be an issue when we try to merge back into v3.3/develop?  I think this indicates that this fork's master is out of sync.

File: resource/quote.go [hunk 2]
```diff
@@ -151,25 +150,21 @@ func (ctx *TpmQuoteContext) readEventLog() error {
...
-	// this was needed to avoid an error in HVS parsing...
-	// 'Current state not START_ELEMENT, END_ELEMENT or ENTITY_REFERENCE'
-	xml := string(eventLogBytes)
-	xml = strings.Replace(xml, " ", "", -1)
-	xml = strings.Replace(xml, "\t", "", -1)
-	xml = strings.Replace(xml, "\n", "", -1)
-
-	eventLog := base64.StdEncoding.EncodeToString([]byte(xml))
+	json := string(eventLogBytes)
+	eventLog := base64.StdEncoding.EncodeToString([]byte(json))
```
Comment (kwtj43) [inline resource/quote.go]: If we don't need to strip out new lines, tabs, etc, then you can get rid of the 'json' variable and just call eventLog := base64.StdEncoding.EncodeToString(eventLogBytes).

File: resource/event_log.go [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return errors.Wrapf(err, "resource/event_log:FetchUefiEventInfo() There was an error getting FileInfo of %s", Tpm2FilePath)
+	}
+
+	// As TPM2 file contains uefi_event_log size from offset 64-67 and uefi_event_log address from offset 68-75, so below check is for uefi_event_log Info in TPM2 file
+	if fileInfo.Size() < 76 {
```
Comment (kwtj43) [inline resource/event_log.go]: Check the binary's 'magic' make sure that is 'TPM2' (to make sure we're not parsing garbage).

File: main.go [hunk 4]
```diff
@@ -194,15 +194,40 @@ func updateMeasureLog() error {
...
+		}
+	}
+
+	err = eventLogInfo.UpdateAppEventLog()
+	if err != nil {
+		log.Errorf("main:updateMeasureLog() Error while updating Application Event Log: %s\n", err.Error())
+	}
+
+	if eventLogInfo.FinalPcrEventLog != nil {
+		eventLogInfo.WriteMeasureLogFile()
```
Comment (kwtj43) [inline main.go]: I think I would have implemented this a little different but will defer to @ddhawal on what should be changed.  My recommendations are...

1. Create a new package/subfolder called eventlog ("resource" is intended for REST endpoints).
2. In that package, define a single "public" EventLogParser interface that has a one method "GetEventLogs() ( []PcrEventLog, error)".  Also expose data public data structures that have json decorations (ex. "PcrData").  
3. Add a free function (NewEventLogParser) that creates an instance of interface. 
4. Keep everything else internal to the package and encapsulate functions like "FetchUefiEventInfo()" and "FetchTxtHeapInfo()".  In other words, "main" would just news up the EventLogParser, calls "GetEventLogs" (returns the []PcrEventLog) and serializes that array to a contants.MeasureLogFilePath.

**Review by kwtj43 — CHANGES_REQUESTED** [review]

File: .gitlab-ci.yml (hunk 1 above)
Comment (mjeyarax) [reply .gitlab-ci.yml]: It was informed that scan job has been added in the pipeline. In v3.2 it was not there so we have referred v3.3/develop and included as we thought it may needed. This change wont be included, we will revert this back.

File: constants/constants.go [hunk 5]
```diff
@@ -25,7 +24,7 @@ const (
 	RamfsDir                   = VarDir + "ramfs/"
 	SystemInfoDir              = VarDir + "system-info/"
 	PlatformInfoFilePath       = SystemInfoDir + "platform-info"
-	MeasureLogFilePath         = VarDir + "measureLog.xml"
+	MeasureLogFilePath         = VarDir + "measureLog.json"
```
Comment (ddhawal) [inline constants/constants.go]: change the name to measure-log.json

File: main.go (hunk 4 above)
Comment (ddhawal) [reply main.go]: I would go with recommendations from @kwtj43 , I had a hard time in understanding how does this come from resource

File: resource/event_log.go [hunk 6]
```diff
@@ -0,0 +1,840 @@
...
+	0x4ff:      "CAP_VALUE",
+}
+
+var eventLogInfo *EventLogInfo
+
+//InitializeEventLogInfo - Function to create and intialize eventLogInfo variable
+func InitializeEventLogInfo() *EventLogInfo {
+	log.Trace("resource/event_log:InitializeEventLogInfo() Entering")
+	defer log.Trace("resource/event_log:InitializeEventLogInfo() Leaving")
+	eventLogInfo := new(EventLogInfo)
```
Comment (ddhawal) [inline resource/event_log.go]: EventLogInfo{} would set all the default values

File: resource/event_log.go [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+	log.Trace("resource/event_log:FetchTxtHeapInfo() Entering")
+	defer log.Trace("resource/event_log:FetchTxtHeapInfo() Leaving")
+
+	txtHeapBaseAddr := make([]byte, Uint64Size)
+	txtHeapSize := make([]byte, Uint64Size)
+	file, err := os.Open(EventLogFilePath)
+	if err != nil {
+		return errors.Wrapf(err, "resource/event_log:FetchTxtHeapInfo() There was an error opening %s", EventLogFilePath)
+	}
+	defer file.Close()
```
Comment (ddhawal) [inline resource/event_log.go]: handle errors in defer

File: resource/event_log.go [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+	log.Trace("resource/event_log:FetchUefiEventInfo() Entering")
+	defer log.Trace("resource/event_log:FetchUefiEventInfo() Leaving")
+
+	uefiBaseAddress := make([]byte, Uint64Size)
+	uefiSize := make([]byte, Uint32Size)
+	file, err := os.Open(Tpm2FilePath)
+	if err != nil {
+		return errors.Wrapf(err, "resource/event_log:FetchUefiEventInfo() There was an error opening %s", Tpm2FilePath)
+	}
+	defer file.Close()
```
Comment (ddhawal) [inline resource/event_log.go]: handle errors in defer

File: resource/event_log.go [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+		pcr := make([]PcrData, tpmlDigestValues.count)
+
+		for hashIndex = 0; hashIndex < int(tpmlDigestValues.count); hashIndex++ {
+
+			var digestSize int
+			var algID uint16
+			binary.Read(buf, binary.LittleEndian, &algID)
+			offset = offset + Uint16Size
+
+			if algID == AlgSHA1 {
```
Comment (ddhawal) [inline resource/event_log.go]: use switch block

File: resource/event_log.go [hunk 10]
```diff
@@ -0,0 +1,840 @@
...
+				digestSize = 48
+				digest = buf.Next(digestSize)
+				offset = offset + int64(digestSize)
+				digestStr := hex.EncodeToString(digest)
+				eventData[hashIndex].Measurement = digestStr
+				pcr[hashIndex].Bank = SHA384
+
+			} else if algID == AlgSHA512 {
+
+				digest = buf.Next(sha512.Size)
```
Comment (ddhawal) [inline resource/event_log.go]: same code repeats in each block can be moved to function

File: resource/event_log.go [hunk 11]
```diff
@@ -0,0 +1,840 @@
...
+			} else {
+				//Handling of 501 Events according to spec.
+				//The first and second  occurrence of 501 events is tb_policy
+				//The third occurrence results in “vmlinuz”.
+				//The fouth occurrence results in “initrd”.
+				//The fifth occurrence results in “asset-tag”.
+				//All other occurrences will be blank.
+
+				if eventTypeStr == Event501 {
+					if event501Index < 2 {
```
Comment (ddhawal) [inline resource/event_log.go]: even negative values? if values are fixed use switch

**Review by ddhawal — CHANGES_REQUESTED** [review]

File: main.go (hunk 4 above)
Comment (ss893) [reply main.go]: @ddhawal 
 We have changed our structure according to Kent's comment.
 Now we have one package/folder called eventlog.
 In this we have two files
 1. event_log.go :
    This file contains only the interface function and public data structure
    i.e  GetEventLogs() and a free function (NewEventLogParser) that creates an instance of interface
 2. collect_event.go: This is a common file to collect all events(txt, uefi, application).       
 or 
 Do you want us to have separate files to collect each txt, uefi and application event ?
 Based on your comment we will create separate file like : collect_txt_event.go, collect_uefi_event.go, collect_application_event.go

File: resource/event_log.go (hunk 6 above)
Comment (kwtj43) [reply resource/event_log.go]: Resolving this comment as the code no longer exists.

File: resource/event_log.go (hunk 8 above)
Comment (kwtj43) [reply resource/event_log.go]: resolving comment -- file.Close() is handled in a defer func() as recommended.

File: resource/event_log.go (hunk 7 above)
Comment (kwtj43) [reply resource/event_log.go]: resolving comment -- file.Close() is handled in a defer func() as recommended.

File: eventlog/collect_txt_event.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+		nextEventLogOffset := (physicalAddress - eventLog.TxtHeapBaseAddr) + uint64(nextRecordOffset)
+		nextEventLogBuffer := bytes.NewBuffer(mmap[nextEventLogOffset:])
+		err = createMeasureLog(nextEventLogBuffer, allocatedEventContainerSize-uint32(nextEventLogOffset))
+		if err != nil {
+			return errors.Wrap(err, "eventlog/collect_txt_event:updateTxtEventLog() Error while parsing nextEventLogOffset data")
+		}
+	}
+
+	// Unmap the /dev/mem buffer
+	err = syscall.Munmap(mmap)
```
Comment (kwtj43) [inline eventlog/collect_txt_event.go]: What will happen if an error condition occurs before this call to Munmap?  Should this be in a "defer" statement after the mmap is created?

File: eventlog/common.go [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+// ParseEventLog - Function to parse event log data from buffer
+func parseEventLog(buf *bytes.Buffer, size uint32) error {
+	log.Trace("eventlog/common:parseEventLog() Entering")
+	defer log.Trace("eventlog/common:parseEventLog() Leaving")
+
+	// Skip TCG_PCR_EVENT(Intel TXT spec. ver. 16.2) log data
+	tcgPcrEvent := tcgPcrEventV1{}
+	err := binary.Read(buf, binary.LittleEndian, &tcgPcrEvent.PcrIndex)
+	if err != nil {
+		return errors.Wrap(err, "eventlog/common:parseEventLog() There is an error reading binary data from buffer")
```
Comment (kwtj43) [inline eventlog/common.go]: Update the error messages so they are not all the same (this helps troubleshooting customer issues).  For example, "There was an error reading the PCR Index from the buffer" or "There was an error reading the Event Type from the buffer").

File: eventlog/common.go [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+
+	tcgPcrEvent2 := tcgPcrEventV2{}
+	tpmlDigestValues := tpmlDigestValue{}
+	var offset int64
+	var err error
+	event501Index := 0
+	for offset = 0; offset < int64(size); {
+		err = binary.Read(buf, binary.LittleEndian, &tcgPcrEvent2.PcrIndex)
+		if err != nil {
+			return errors.Wrap(err, "eventlog/common:createMeasureLog() There is an error reading binary data from buffer")
```
Comment (kwtj43) [inline eventlog/common.go]: Again, differentiate error messages.

File: eventlog/common.go [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+	0x414:      "NV_INFO_HASH",
+	0x416:      "EVTYPE_KM_HASH",
+	0x417:      "EVTYPE_BPM_HASH",
+	0x418:      "EVTYPE_KM_INFO_HASH",
+	0x419:      "EVTYPE_BPM_INFO_HASH",
+	0x41a:      "EVTYPE_BOOT_POL_HASH",
+	0x4ff:      "CAP_VALUE",
+}
+
+var evLogInfo *eventLogInfo
```
Comment (kwtj43) [inline eventlog/common.go]: Avoid global variables where possible -- I don't see a need for a singleton here.  I would suggest creating an instance of eventLogInfo in event_log.go::GetEventLog() and removing initializeEventLogInfo() and getEventLogInfo().

**Review by kwtj43 — CHANGES_REQUESTED** [review]

**Comment by kwtj43** [comment]
Thanks for resolving the conflicts -- feel free to merge.

**Review by kwtj43 — APPROVED** [review]
Thanks fore resolving the comments.

File: eventlog/event_log.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+// GetEventLogs extracts the eventlogs data and returns these for serialization of array to constants.MeasureLogFilePath
+func (evtLogFile *EventLogFiles) GetEventLogs() ([]PcrEventLog, error) {
+	log.Trace("eventlog/event_log:GetEventLogs() Entering")
+	defer log.Trace("eventlog/event_log:GetEventLogs() Leaving")
+
+	eventLog := eventLogInfo{}
+	err := eventLog.fetchUefiEventInfo(evtLogFile.Tpm2FilePath)
```
Comment (ddhawal) [inline eventlog/event_log.go]: fetchUefiEventInfo/fetchTxtHeapInfo/updateUefiEventLog does not seem to be related to eventLog, data can be collected using these functions, and then they can return this data. This event log data can then be assigned to eventLog here instead of data getting updated in the individual functions.

**Review by ddhawal — APPROVED** [review]
